Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcContentServer.cs

```diff
@@ -153,97 +156,174 @@ public async Task<GetStatsResponse> GetStatsAsync(GetStatsRequest request, Cance
-                }
+                    // Figure out response headers.
+                    CopyCompression compression = CopyCompression.None;
+                    Metadata headers = new Metadata();
+                    headers.Add("ContentHash", hash.ToString());
+                    switch (result.Code)
+                    {
+                        case OpenStreamResult.ResultCode.ContentNotFound:
+                            headers.Add("Exception", "ContentNotFound");
+                            headers.Add("Message", $"Requested content {hash} not found.");
+                            break;
+                        case OpenStreamResult.ResultCode.Error:
+                            Debug.Assert(result.Exception != null);
+                            headers.Add("Exception", result.Exception.GetType().Name);
+                            headers.Add("Message", result.Exception.Message);
+                            break;
+                        case OpenStreamResult.ResultCode.Success:
+                            Debug.Assert(result.Stream != null);
+                            long size = result.Stream.Length;
+                            headers.Add("FileSize", size.ToString());
+                            if ((request.Compression == CopyCompression.Gzip) && (size > ContentStore.Grpc.CopyConstants.DefaultBufferSize)) {
+                                compression = CopyCompression.Gzip;
+                            }
+                            headers.Add("Compression", compression.ToString());
+                            headers.Add("ChunkSize", ContentStore.Grpc.CopyConstants.DefaultBufferSize.ToString());
+                            break;
+                    }
 
+                    // Send the response headers.
+                    await context.WriteResponseHeadersAsync(headers);
 
-                if (!(cache is IStreamStore copyStore))
-                {
-                    await responseStream.WriteAsync(
-                        new CopyFileResponse
-                        {
-                            Header = new ResponseHeader(startTime, false, (int)CopyFileResult.ResultCode.SourcePathError, $"'{request.Drive}' does not support copying."),
-                        });
-                    return;
-                }
-
-                var openStreamResult = await copyStore.StreamContentAsync(cacheContext, request.ContentHash.ToContentHash((HashType)request.HashType));
-                if (openStreamResult.Succeeded)
-                {
-                    await StreamContentAsync(openStreamResult, responseStream, startTime);
-                }
-                else
-                {
-                    await responseStream.WriteAsync(
-                        new CopyFileResponse
+                    // Send the content.
+                    if (result.Succeeded)
+                    {
+                        byte[] buffer = new byte[ContentStore.Grpc.CopyConstants.DefaultBufferSize];

```

Comment:
question: do we want to pull this? Or the lifetime is more complicated and not a scoped based?

Reply:
:+1

---
In reply to: [272773954](https://github.com/Microsoft/BuildXL/pull/98#discussion_r272773954) [](ancestors = 272773954)

Reply:
I assume s/pull/pool. Yes, eventually, but I didn't want to deal with that here.